common/src/main/java/com/wynntils/features/map/WorldMarkersFeature.java

```diff
@@ -0,0 +1,539 @@
+import com.wynntils.core.persisted.config.Config;
+import com.wynntils.core.persisted.config.ConfigCategory;
+import com.wynntils.core.text.StyledText;
+import com.wynntils.mc.event.RenderEvent;
+import com.wynntils.mc.event.RenderTileLevelLastEvent;
+import com.wynntils.mc.event.TickEvent;
+import com.wynntils.services.map.pois.WaypointPoi;
+import com.wynntils.services.mapdata.attributes.type.MapIcon;
+import com.wynntils.services.mapdata.attributes.type.ResolvedMapAttributes;
+import com.wynntils.services.mapdata.attributes.type.ResolvedMarkerOptions;
+import com.wynntils.services.mapdata.type.MapLocation;
+import com.wynntils.utils.MathUtils;
+import com.wynntils.utils.colors.CommonColors;
+import com.wynntils.utils.colors.CustomColor;
+import com.wynntils.utils.mc.McUtils;
+import com.wynntils.utils.mc.type.Location;
+import com.wynntils.utils.render.FontRenderer;
+import com.wynntils.utils.render.RenderUtils;
+import com.wynntils.utils.render.Texture;
+import com.wynntils.utils.render.type.HorizontalAlignment;
+import com.wynntils.utils.render.type.TextShadow;
+import com.wynntils.utils.render.type.VerticalAlignment;
+import com.wynntils.utils.type.Pair;
+import java.util.ArrayList;
+import java.util.List;
+import net.minecraft.client.Camera;
+import net.minecraft.client.renderer.MultiBufferSource;
+import net.minecraft.client.renderer.blockentity.BeaconRenderer;
+import net.minecraft.core.Position;
+import net.minecraft.world.phys.Vec2;
+import net.minecraft.world.phys.Vec3;
+import net.neoforged.bus.api.SubscribeEvent;
+import org.joml.Matrix4f;
+import org.joml.Quaternionf;
+import org.joml.Vector3d;
+import org.joml.Vector3f;
+import org.joml.Vector4f;
+
+@ConfigCategory(Category.MAP)
+public class WorldMarkersFeature extends Feature {
+    private static final float MINIMUM_RENDER_VISIBILITY = 0.1f;
+    private static final int RAINBOW_CHANGE_RATE = 10;
+
+    private static final MultiBufferSource.BufferSource BUFFER_SOURCE =
+            MultiBufferSource.immediate(new ByteBufferBuilder(256));
+
+    private static final WaypointPoi DUMMY_WAYPOINT = new WaypointPoi(() -> null, "");
+
+    @Persisted
+    public final Config<Float> textBackgroundOpacity = new Config<>(0.2f);
+
+    @Persisted
+    public final Config<Float> textRenderScale = new Config<>(1.0f);
+
+    @Persisted
+    public final Config<Float> iconRenderScale = new Config<>(1.0f);
+
+    @Persisted
+    public final Config<TextShadow> textShadow = new Config<>(TextShadow.NONE);
+
+    @Persisted
+    public final Config<Float> bottomBoundingDistance = new Config<>(100f);
+
+    @Persisted
+    public final Config<Float> topBoundingDistance = new Config<>(40f);
+
+    @Persisted
+    public final Config<Float> horizontalBoundingDistance = new Config<>(30f);
+
+    @Persisted
+    public final Config<Integer> maxMarkerDistance = new Config<>(5000);
+
+    @Persisted
+    public final Config<CustomColor> markerBeaconColor = new Config<>(CommonColors.RED);
```

Comment:
But this surely conflicts with the map attributes, right?

Reply:
Can we have that as beacon color in our default base attribute instead?

Reply:
Yes, sure, I haven't thought about that.

Reply:
In fact, what if I just remove the default color, and leave it up to the user for it to be "required" for rendering? I think it makes sense that the user has to define the color, the min max ranges, etc.

Reply:
What do you think about this, @magicus?